velox/functions/lib/Re2Functions.h

```diff
@@ -255,55 +272,79 @@ template <
       const std::vector<TypePtr>& /*inputTypes*/,
       const core::QueryConfig& config,
       const arg_type<Varchar>* /*string*/,
       const arg_type<Varchar>* pattern,
       const arg_type<Varchar>* replacement) {
-    VELOX_USER_CHECK(
-        pattern != nullptr, "Pattern of regexp_replace must be constant.");
-    VELOX_USER_CHECK(
-        replacement != nullptr,
-        "Replacement sequence of regexp_replace must be constant.");
-
-    auto processedPattern = prepareRegexpPattern(*pattern);
-
-    re_.emplace(processedPattern, RE2::Quiet);
-    if (UNLIKELY(!re_->ok())) {
-      VELOX_USER_FAIL(
-          "Invalid regular expression {}: {}.", processedPattern, re_->error());
+    if (pattern != nullptr) {
+      const auto processedPattern = prepareRegexpPattern(*pattern);
+      re_.emplace(processedPattern, RE2::Quiet);
+      VELOX_USER_CHECK(
+          re_->ok(),
+          "Invalid regular expression {}: {}.",
+          processedPattern,
+          re_->error());
     }
 
-    processedReplacement_ = prepareRegexpReplacement(*re_, *replacement);
+    if (replacement != nullptr) {
+      VELOX_USER_CHECK_NOT_NULL(
+          pattern, "Pattern of regexp_replace must be constant");
+
+      ensureProcessedReplacement(re_.value(), *replacement);
+      constantReplacement_ = true;
+    }
   }
 
   FOLLY_ALWAYS_INLINE void initialize(
       const std::vector<TypePtr>& inputTypes,
       const core::QueryConfig& config,
       const arg_type<Varchar>* string,
       const arg_type<Varchar>* pattern) {
-    StringView emptyReplacement;
-
-    initialize(inputTypes, config, string, pattern, &emptyReplacement);
+    initialize(inputTypes, config, string, pattern, nullptr);
   }
 
-  FOLLY_ALWAYS_INLINE bool call(
+  FOLLY_ALWAYS_INLINE void call(
       out_type<Varchar>& out,
       const arg_type<Varchar>& string,
-      const arg_type<Varchar>& /*pattern*/,
-      const arg_type<Varchar>& /*replacement*/ = StringView{}) {
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replacement = StringView{}) {
+    auto& re = ensurePattern(pattern);
+    const auto& processedReplacement =
+        ensureProcessedReplacement(re, replacement);
+
     result_.assign(string.data(), string.size());
-    RE2::GlobalReplace(&result_, *re_, processedReplacement_);
+    RE2::GlobalReplace(&result_, re, processedReplacement);
 
     UDFOutputString::assign(out, result_);
+  }
+
+ private:
+  RE2& ensurePattern(const arg_type<Varchar>& pattern) {
+    if (!re_.has_value()) {
+      auto processedPattern = prepareRegexpPattern(pattern);
+      return *cache_.findOrCompile(StringView(processedPattern));
+    } else {
+      return re_.value();
+    }
+  }
 
-    return true;
+  const std::string& ensureProcessedReplacement(
+      RE2& re,
+      const arg_type<Varchar>& replacement) {
+    if (!constantReplacement_) {
+      processedReplacement_ = prepareRegexpReplacement(re, replacement);
+    }
```

Comment:
IIUC, re_.has_value() means the function call uses constant pattern. constantReplacement_ means the function call uses constant replacement. Could we also make processedReplacement_.has_value imply constant replacement instead of using the constantReplacement_ flag? I feel this would make the code for handling pattern and replacement more consistent.

Reply:
> Could we also make processedReplacement_.has_value imply constant replacement instead of using the constantReplacement_ flag?

We would then need another string to store non-constant replacements. That's doable, but seems a bit wasteful. Let me know if you want me to make that change.

Reply:
That's OK. Then, could we instead add a comment that re_ is used only when there is a constant pattern?